libraries/integration/Microsoft.Bot.Builder.ApplicationInsights.Core/TelemetryBotIdInitializer.cs

```diff
@@ -16,44 +15,51 @@ namespace Microsoft.Bot.Builder.ApplicationInsights.Core
     /// </summary>
     public class TelemetryBotIdInitializer : ITelemetryInitializer
     {
-        private readonly IMemoryCache _memoryCache;
+        public static readonly string BotActivityKey = "BotActivity";
         private readonly IHttpContextAccessor _httpContextAccessor;
 
-        public TelemetryBotIdInitializer(IHttpContextAccessor httpContextAccessor, IMemoryCache memoryCache)
+        public TelemetryBotIdInitializer(IHttpContextAccessor httpContextAccessor)
         {
-            _memoryCache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
             _httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
         }
 
         public void Initialize(ITelemetry telemetry)
         {
-            var context = _httpContextAccessor.HttpContext;
-            if (context != null && (telemetry is RequestTelemetry || telemetry is EventTelemetry))
+            if (telemetry == null)
             {
-                // can't read from the request body at this point, as the
-                // request stream has already been disposed.
-                var body = _memoryCache.Get(context.TraceIdentifier) as JObject;
-                if (body != null)
+                return;
+            }
+
+            var httpContext = _httpContextAccessor.HttpContext;
+            var items = httpContext?.Items;
+
+            if (items != null)
+            {
+                if ((telemetry is RequestTelemetry || telemetry is EventTelemetry) && items.ContainsKey(BotActivityKey))
                 {
-                    var userId = (string)body["from"]?["id"];
-                    var channelId = (string)body["channelId"];
-                    var conversationId = (string)body["conversation"]?["id"];
+                    var body = items[BotActivityKey] as JObject;
+                    if (body != null)
+                    {
+                        var userId = (string)body["from"]?["id"];
+                        var channelId = (string)body["channelId"];
+                        var conversationId = (string)body["conversation"]?["id"];
 
-                    // Set the user id on the Application Insights telemetry item.
-                    telemetry.Context.User.Id = channelId + userId;
+                        // Set the user id on the Application Insights telemetry item.
+                        telemetry.Context.User.Id = channelId + userId;
```

Comment:
You use the `telemetry.Context` property multiple times in a row here. Suggest putting in a variable. Even though the argument could be made that the JIT is able to optimize the call to the property accessor away, it's still noisier code than just introducing a variable and reusing it.

Reply:
Fixed